### PR TITLE
CLOUDSTACK-8819: Added QCOW2 virtual size checking for S3.

### PR DESCRIPTION
**Pull request by borisroman**
- Cleaned up S3TemplateDownloader
- Created static QCOW2 utils class.
- Reformatted some parts of DownloadManagerImpl

### CODE REVIEW

**Comment by borisroman** [comment]
Tested manually. Virtual size is correctly returned now.

File: utils/src/main/java/com/cloud/utils/storage/QCOW2Utils.java [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+     * @return The virtual size of the QCOW2 object.
+     */
+    public static long getVirtualSize(InputStream inputStream) throws IOException {
+        byte[] b = new byte[8];
+
+        if (inputStream.skip(VIRTUALSIZE_HEADER_LOCATION) != VIRTUALSIZE_HEADER_LOCATION) {
+            throw new IOException("Unable to skip to the virtual size header");
+        }
+
+        if (inputStream.read(b) != 8) {
```
Comment (wido) [inline utils/src/main/java/com/cloud/utils/storage/QCOW2Utils.java]: Should you use VIRTUALSIZE_HEADER_LENGTH here?

Comment (borisroman) [inline utils/src/main/java/com/cloud/utils/storage/QCOW2Utils.java]: True that.

**Comment by wido** [comment]
LGTM

**Comment by bhaisaab** [comment]
LGTM

**Comment by bhaisaab** [comment]
@borisroman looks good, can you increase code coverage by adding tests for the utils and core changes?

**Comment by borisroman** [comment]
@bhaisaab Will do! Please wait with merging, will commit them to this PR.

**Comment by bhaisaab** [comment]
Thanks @borisroman

**Comment by borisroman** [comment]
@karuturi Fixed the findbug issue. Was due to reformatting of one method.

File: services/secondary-storage/server/src/org/apache/cloudstack/storage/template/DownloadManagerImpl.java [hunk 2]
```diff
@@ -382,7 +424,7 @@ private String postDownload(String jobId) {
         if (extension.equals("iso")) {
             templateName = jobs.get(jobId).getTmpltName().trim().replace(" ", "_");
         } else {
-            templateName = java.util.UUID.nameUUIDFromBytes((jobs.get(jobId).getTmpltName() + System.currentTimeMillis()).getBytes()).toString();
+            templateName = java.util.UUID.nameUUIDFromBytes((jobs.get(jobId).getTmpltName() + System.currentTimeMillis()).getBytes(Charset.forName("UTF-8"))).toString();
```
Comment (karuturi) [inline services/secondary-storage/server/src/org/apache/cloudstack/storage/template/DownloadManagerImpl.java]: Can you use StringUtils.getPreferredCharset()?

**Pull request merged**